### PR TITLE
Avoid recursion in Entity::get() when calling the API within `civi.api4.createRequest` event subscribers

### PR DESCRIPTION
**Pull request by jensschuppe**
Overview
----------------------------------------
Follow-up to #21828 for avoiding recursion when subscribing to the new `civi.api4.entityTypes` event and calling the API within the subscriber

Before
----------------------------------------
There's a recursion when the new `civi.api4.entityTypes` event is being implemented and the API is being called from within the event subscriber. The code path always drops in at https://github.com/civicrm/civicrm-core/blob/ca09f9073ec01ba0a67e1c435213dea878069299/CRM/Core/BAO/CustomField.php#L342 which does this for every `$customDataType` to check for whether they're a contact sub type. This is resulting in `\Civi\Api4\Action\Entity\Get::getRecords()` being called eventually, where the `civi.api4.entityTypes` event was added, jumping into each subscriber → Recursion!

After
----------------------------------------
No recursion.

Technical Details
----------------------------------------
Instead of trying to get all contact sub types using the API (which `CRM_Contact_BAO_ContactType::subTypes()` does), use `CRM_Contact_BAO_ContactType::getBasicType()` for deciding whether the currently processed API entity is a contact sub type.

Comments
----------------------------------------
That seems like a bogus workaround, but `CRM_Contact_BAO_ContactType::getBasicType()` does not call the API for retrieving its result. I have no idea how else to solve that, so this should be considered a rather hacky fix.

Pinging @colemanw for a review.

### CODE REVIEW

File: CRM/Contact/BAO/ContactType.php [hunk 1]
```diff
@@ -871,14 +869,17 @@ protected static function getAllContactTypes() {
     $cacheKey = 'all_' . $GLOBALS['tsLocale'];
     $contactTypes = $cache->get($cacheKey);
     if ($contactTypes === NULL) {
-      $contactTypes = (array) ContactType::get(FALSE)
-        ->setSelect(['id', 'name', 'label', 'description', 'is_active', 'is_reserved', 'image_URL', 'parent_id', 'parent_id:name', 'parent_id:label'])
-        ->execute()->indexBy('name');
-
+      $query = CRM_Utils_SQL_Select::from('civicrm_contact_type')
+        ->select(['id', 'name', 'label', 'description', 'is_active', 'is_reserved', 'image_URL', 'parent_id']);
```
Comment (colemanw) [inline CRM/Contact/BAO/ContactType.php:873]: This line can cause problems during upgrades by trying to select a column that hasn't been added to the schema yet. I think it's safer to remove this line (which will default to `SELECT *`).

File: CRM/Contact/BAO/ContactType.php [hunk 2]
```diff
@@ -871,14 +869,17 @@ protected static function getAllContactTypes() {
...
+      $contactTypes = array_column($dao->fetchAll(), NULL, 'name');
+      $name_options = self::buildOptions('parent_id', 'validate');
+      $label_options = self::buildOptions('parent_id', 'get');
       foreach ($contactTypes as $id => $contactType) {
-        $contactTypes[$id]['parent'] = $contactType['parent_id:name'];
-        $contactTypes[$id]['parent_label'] = $contactType['parent_id:label'];
-        unset($contactTypes[$id]['parent_id:name'], $contactTypes[$id]['parent_id:label']);
+        if ($contactType['parent_id']) {
+          $contactTypes[$id]['parent'] = $name_options[$contactType['parent_id']];
+          $contactTypes[$id]['parent_label'] = $label_options[$contactType['parent_id']];
```
Comment (colemanw) [inline CRM/Contact/BAO/ContactType.php:881]: To fix the failing test, add something to the loop to cast some values to the correct type (int or bool)

File: CRM/Contact/BAO/ContactType.php (hunk 1 above)
Comment (jensschuppe) [reply CRM/Contact/BAO/ContactType.php]: Removed the select in 4f4c08155595d678386a99cff3f13540861ecf4c. The tests will have to be adapted though, once there are changes to the schema.

**Comment by colemanw** [comment]
I've opened #21851 with rebasing and additional cleanup.

**Pull request closed without merging**